pkg/util/clusteroperator/status_manager.go

```diff
@@ -0,0 +1,184 @@
+		logrus.Infof("created clusteroperator: %#v", co)
+	} else {
+		if !ConditionsEqual(oldConditions, co.Status.Conditions) {
+			err = status.client.Status().Update(context.TODO(), co)
+			if err != nil {
+				log.Printf("failed to update clusteroperator %s: %v", co.Name, err)
+				return fmt.Errorf("failed to update clusteroperator %s: %v", co.Name, err)
+			}
+			logrus.Infof("updated clusteroperator: %#v", co)
+ 		}
+	}
+	return nil
+}
+
+// SetConfigFailure marks the operator as Failing due to a configuration problem. Attempts
+// to mark the operator as Progressing or Available will be ignored until SetConfigSuccess
+// is called to clear the config error.
+func (status *StatusManager) SetConfigFailing(reason string, err error) error {
+	status.configFailure = true
+	return status.SetFailing(reason, err)
+}
+
+// SetConfigSuccess marks the operator as having a valid configuration, allowing it
+// to be set Progressing or Available.
+func (status *StatusManager) SetConfigSuccess() error {
+	status.configFailure = false
+	return status.Set(
+		&configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.OperatorFailing,
+			Status:  configv1.ConditionFalse,
+		},
+	)
+}
+
+// SetFailing marks the operator as Failing, with the given reason and error message.
+// Unlike with SetConfigFailing, this failure is not persistent.
+func (status *StatusManager) SetFailing(reason string, err error) error {
+	return status.Set(
+		&configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.OperatorFailing,
+			Status:  configv1.ConditionTrue,
+			Reason:  reason,
+			Message: err.Error(),
+		},
+		&configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.OperatorProgressing,
+			Status:  configv1.ConditionFalse,
+			Reason:  "Failing",
+		},
+		&configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.OperatorAvailable,
+			Status:  configv1.ConditionFalse,
+			Reason:  "Failing",
+		},
+	)
+}
+
+// SetFromDaemonSets sets the operator status to Failing, Progressing, or Available, based
+// on the current status of the indicated DaemonSets. However, this is a no-op if the
+// StatusManager is currently marked as failing due to a configuration error.
+func (status *StatusManager) SetFromDaemonSets(daemonSets []types.NamespacedName) error {
+	if status.configFailure {
+		return nil
+	}
+
+	progressing := []string{}
+
+	for _, dsName := range daemonSets {
+		ns := &corev1.Namespace{}
+		if err := status.client.Get(context.TODO(), types.NamespacedName{Name: dsName.Namespace}, ns); err != nil {
+			if errors.IsNotFound(err) {
+				return status.SetFailing("NoNamespace", fmt.Errorf("Namespace %q does not exist", dsName.Namespace))
+			} else {
+				return status.SetFailing("InternalError", err)
+			}
+		}
+
+		ds := &appsv1.DaemonSet{}
+		if err := status.client.Get(context.TODO(), dsName, ds); err != nil {
+			if errors.IsNotFound(err) {
+				return status.SetFailing("NoDaemonSet", fmt.Errorf("DaemonSet %q does not exist", dsName.String()))
```

Comment:
Is it useful to know if more than one daemonset is missing?

Reply:
yeah, probably

Reply:
Actually, probably not; this would normally just be a transient failure during the first few moments of the cluster's existence, and no one will ever actually see it. (And if we're going to combine errors then what if we're missing the Namespace for one DaemonSet and the DaemonSet for another, etc?)